File: packages/schema/signal.py
"""Python mirror of signal.ts — Pydantic models for vision/audio frames + windowed bundle."""

from __future__ import annotations

from typing import Optional, List

from pydantic import BaseModel, Field


class VisionFrame(BaseModel):
    t: float = Field(ge=0)
    gaze_fixation_ratio: float = Field(ge=0, le=1)
    posture_sway: float = Field(ge=0)
    shoulder_tilt: float
    expression_diversity: float = Field(ge=0)
    hand_gesture_freq: float = Field(ge=0)
    # Peak wrist speed (image-normalized units/sec) in the last ~1s rolling window.
    # Spikes signal sudden burst gestures; sustained spikes become GESTURE_EXCESSIVE.
    hand_velocity_max: float = Field(default=0.0, ge=0)
    # Head pose decomposed from MediaPipe face transformation matrix (Euler angles, degrees).
    # +pitch = looking down (chin to chest), +yaw = head turned to subject's left,
    # +roll = head tilted toward subject's right shoulder.
    head_pitch_deg: float = 0.0
    head_yaw_deg: float = 0.0
    head_roll_deg: float = 0.0
    # True when the user is propping head on hand (wrist landmark close to face bounds).
    chin_on_hand: bool = False
    # Mouth-open ratio (jawOpen blendshape, 0..1). Proxy for whether user is speaking
    # at this instant; combined with audio RMS it lets us distinguish silent-mouth-open
    # vs. talking.
    mouth_open: float = Field(default=0.0, ge=0, le=1)

    # ── Domain-expansion signals (드라이브: 소개팅·고객응대·면접 등 시나리오 평가 확장) ──
    # Smile blendshape (avg of mouthSmileLeft/Right, 0..1). Warmth signal —
    # 소개팅/고객응대 시나리오에서 강한 가중치.
    smile_intensity: float = Field(default=0.0, ge=0, le=1)
    # Wrist inside upper face region (cheek/nose/forehead) but NOT chin — covers
    # nose-touching / hair-pushing / forehead-rubbing (anxiety signals).
    face_touch_other: bool = False
    # 0..1 — both wrists close together with small repetitive motion = 안절부절 fidget.
    hand_fidget_score: float = Field(default=0.0, ge=0)
    # Wrist direction reversals per second in last ~2s. High = 이중동작/망설임.
    motion_reversal_rate: float = Field(default=0.0, ge=0)
    # Head-yaw stddev in last ~2s (degrees). High = 시선 좌우 산만.
    gaze_yaw_sway: float = Field(default=0.0, ge=0)
    # Frame-to-frame blendshape change rate. High = 풍부한 표정 변화; low = 무표정.
    expression_change_rate: float = Field(default=0.0, ge=0)


class SttWord(BaseModel):
    t_start: float
    t_end: float
    word: str
    prob: Optional[float] = None


class SttSegment(BaseModel):
    t_start: float = Field(ge=0)
    t_end: float = Field(ge=0)
    text: str
    words: List[SttWord] = Field(default_factory=list)
    is_final: bool = True


class ProsodyFrame(BaseModel):
    t_start: float = Field(ge=0)
    t_end: float = Field(ge=0)
    wpm: float = Field(default=0.0, ge=0)
    # Number of STT words/eojeols in this frame. This lets the aggregator compute
    # session-level WPM from actual counts instead of averaging per-segment rates.
    word_count: int = Field(default=0, ge=0)
    filler_count: int = Field(default=0, ge=0)
    filler_terms: List[str] = Field(default_factory=list)
    pause_seconds: float = Field(default=0.0, ge=0)
    # Continuous silence duration ending in this window. Browser-side RMS detector
    # populates this even when STT is not yet running.
    silence_seconds: float = Field(default=0.0, ge=0)
    f0_variance: Optional[float] = Field(default=None, ge=0)
    rms_mean: Optional[float] = Field(default=None, ge=0)
    # ── Phase 2 prosody (audio-pipeline /analyze populates these per STT segment) ──
    # Pitch standard deviation in semitones — canonical monotone metric.
    # Low values = 단조로운 톤, high values = 표현력 있는 억양.
    pitch_sd_semitones: Optional[float] = Field(default=None, ge=0)
    # Pitch range (max-min) in semitones — expressiveness ceiling.
    pitch_range_semitones: Optional[float] = Field(default=None, ge=0)
    # Coefficient of variation of intensity envelope — 강약 다이내믹.
    # Low = 평평한 음량, high = 다양한 강조.
    intensity_cv: Optional[float] = Field(default=None, ge=0)
    # Ratio of energy in the last 0.5s to segment mean. <0.7 = 말끝 흐림.
    end_energy_drop: Optional[float] = Field(default=None, ge=0)
    # Mean Whisper word confidence — proxy for pronunciation clarity.
    # 0..1; <0.6 often means slurred/unclear articulation.
    articulation_proxy: Optional[float] = Field(default=None, ge=0, le=1)


class SignalWindow(BaseModel):
    session_id: str
    t_start: float = Field(ge=0)
    t_end: float = Field(ge=0)
    vision: Optional[VisionFrame] = None  # mean of frames in window
    prosody: Optional[ProsodyFrame] = None
    transcript: str = ""

File: packages/schema/signal.ts
import { z } from 'zod';

// One 5fps vision-signal frame emitted by the browser client.
// All values relative to the recording start (seconds).
export const VisionFrameSchema = z.object({
  t: z.number().min(0),
  gaze_fixation_ratio: z.number().min(0).max(1),
  posture_sway: z.number().min(0),
  shoulder_tilt: z.number(),
  expression_diversity: z.number().min(0),
  hand_gesture_freq: z.number().min(0),
  head_pitch_deg: z.number().default(0),
  head_yaw_deg: z.number().default(0),
  head_roll_deg: z.number().default(0),
  chin_on_hand: z.boolean().default(false),
  mouth_open: z.number().min(0).max(1).default(0),
});
export type VisionFrame = z.infer<typeof VisionFrameSchema>;

// One STT segment from the live audio pipeline.
export const SttSegmentSchema = z.object({
  t_start: z.number().min(0),
  t_end: z.number().min(0),
  text: z.string(),
  words: z
    .array(
      z.object({
        t_start: z.number(),
        t_end: z.number(),
        word: z.string(),
        prob: z.number().optional(),
      }),
    )
    .default([]),
  is_final: z.boolean().default(true),
});
export type SttSegment = z.infer<typeof SttSegmentSchema>;

// Per-window prosody aggregate from the audio pipeline OR a synthetic frame the
// browser pushes for silence-only detection while STT is not yet running.
export const ProsodyFrameSchema = z.object({
  t_start: z.number().min(0),
  t_end: z.number().min(0),
  wpm: z.number().min(0).default(0),
  word_count: z.number().int().min(0).default(0),
  filler_count: z.number().int().min(0).default(0),
  filler_terms: z.array(z.string()).default([]),
  pause_seconds: z.number().min(0).default(0),
  silence_seconds: z.number().min(0).default(0),
  f0_variance: z.number().min(0).optional(),
  rms_mean: z.number().min(0).optional(),
});
export type ProsodyFrame = z.infer<typeof ProsodyFrameSchema>;

// Aggregated 5-second window (output of aggregator before forwarding to coach /live).
export const SignalWindowSchema = z.object({
  session_id: z.string(),
  t_start: z.number().min(0),
  t_end: z.number().min(0),
  vision: VisionFrameSchema.partial().nullable(), // mean of frames in window
  prosody: ProsodyFrameSchema.nullable(),
  transcript: z.string(), // concatenated STT text within the window
});
export type SignalWindow = z.infer<typeof SignalWindowSchema>;
